### PR TITLE
feat(Cluster): rework cluster page

### LINKED ISSUE #1389: Cluster overview - development issue
Design #1257 is approved, stick to version 3.
Screenshot 2024-10-02 at 16 41 14

### PR DESCRIPTION
**Pull request by Raubzeug**
closes #1389 
[stand](https://nda.ya.ru/t/SaCc5lBX78xeSS)

### Test Status: <span style="color: green;">✅ PASSED</span>
📊 [Full Report](https://ydb-platform.github.io/ydb-embedded-ui/1473/)

| Total | Passed | Failed | Flaky | Skipped |
|:-----:|:------:|:------:|:-----:|:-------:|
| 134 | 134 | 0 | 0 | 0 |

### Bundle Size: 🔺
Current: 79.02 MB | Main: 78.97 MB
Diff: +0.05 MB (0.06%)

⚠️ Bundle size increased. Please review.

<details>
<summary>ℹ️ CI Information</summary>

- Test recordings for failed tests are available in the full report.
- Bundle size is measured for the entire 'dist' directory.
- 📊 indicates links to detailed reports.
- 🔺 indicates increase, 🔽 decrease, and ✅ no change in bundle size.
</details>

### CODE REVIEW

File: src/components/DoughnutMetrics/DoughnutMetrics.tsx [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+    if (fillWidth > 50) {
+        gradientFill = 'var(--doughnut-color)';
+        filledDegrees = fillWidth * 3.6 + 90;
+    }
+    const gradientDegrees = filledDegrees;
+    return (
+        <div className={b(null, className)}>
+            <div
+                style={{
```
Comment (astandrik) [inline src/components/DoughnutMetrics/DoughnutMetrics.tsx:52]: Would you try to make it using subclasses?

File: src/containers/Cluster/ClusterDashboard/ClusterDashboard.tsx [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        const {MemoryTotal, MemoryUsed} = cluster;
+        if (valueIsDefined(MemoryTotal) && valueIsDefined(MemoryUsed)) {
+            metricsCards.push(
+                <ClusterMetricsMemory key="memory" value={MemoryUsed} capacity={MemoryTotal} />,
+            );
+        }
+        return metricsCards;
+    };
+
+    const getCards = () => {
```
Comment (astandrik) [inline src/containers/Cluster/ClusterDashboard/ClusterDashboard.tsx:75]: imho It would be better to make it distinct component to make the main component more readable

File: src/containers/Cluster/ClusterDashboard/ClusterDashboard.tsx [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+interface ClusterDashboardProps {
+    cluster: TClusterInfo;
+    groupStats?: ClusterGroupsStats;
+    loading?: boolean;
+}
+
+export function ClusterDashboard({cluster, groupStats = {}, loading}: ClusterDashboardProps) {
+    const getDoughnuts = () => {
+        if (loading) {
+            return Array.from('123').map((el) => <ClusterMetricsCardSkeleton key={el} />);
```
[the remark this answers is not in the document]
Comment (astandrik) [reply src/containers/Cluster/ClusterDashboard/ClusterDashboard.tsx:49]: or make ClusterMetricsCardSkeleton to do the thing by itself and do just

return <ClusterMetricsCardSkeleton />

File: src/containers/Cluster/ClusterDashboard/components/ClusterMetricsCard.tsx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+export function ClusterMetricsCardDoughnut({
+    title,
+    children,
+    size,
+    ...rest
+}: ClusterMetricsDougnutCardProps) {
+    return (
+        <ClusterMetricsCard title={title} size={size}>
+            <DoughnutMetrics {...rest} className={b('doughtnut')}>
```
Comment (astandrik) [inline src/containers/Cluster/ClusterDashboard/components/ClusterMetricsCard.tsx:48]: doughtnut -> doughnut

File: src/containers/Cluster/ClusterDashboard/utils.tsx [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+const rolesToShow = ['storage', 'tenant'];
+
+export function getNodesRolesInfo(cluster: TClusterInfo) {
+    const nodesRoles: React.ReactNode[] = [];
+    if (isClusterInfoV2(cluster) && cluster.MapNodeRoles) {
+        for (const [role, count] of Object.entries(cluster.MapNodeRoles)) {
+            if (rolesToShow.includes(role.toLowerCase())) {
+                nodesRoles.push(
+                    <React.Fragment key={role}>
+                        {role}: {formatNumber(count)}
```
Comment (astandrik) [inline src/containers/Cluster/ClusterDashboard/utils.tsx:61]: letsmake this parameterized i18n:

"key": "{{role}}: {{value}}"

with usage

i18n("key", {role, value: formatNumber(count)})

File: src/containers/Cluster/ClusterDashboard/utils.tsx [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    const result: React.ReactNode[] = [];
+
+    Object.entries(groupStats).forEach(([storageType, stats]) => {
+        Object.values(stats).forEach((erasureStats) => {
+            result.push(
+                <DiskGroupsErasureStats
+                    stats={erasureStats}
+                    key={`${storageType}|${erasureStats.erasure}`}
+                >
+                    {storageType}: {formatNumber(erasureStats.createdGroups)} /{' '}
```
Comment (astandrik) [inline src/containers/Cluster/ClusterDashboard/utils.tsx:80]: same here

File: src/utils/dataFormatters/dataFormatters.ts [hunk 7]
```diff
@@ -63,13 +68,28 @@ export const formatStorageValues = (value?: number, total?: number, size?: Bytes
...
+    let formattedValue = formatBytesCustom({
         value,
         withSizeLabel: valueWithSizeLabel,
         size: size || calculatedSize,
         precision: valuePrecision,
     });
-    const formattedTotal = formatBytesCustom({value: total, size: size || calculatedSize});
+    if (value && value > 0) {
+        while (formattedValue === '0') {
+            valuePrecision += 1;
```
Comment (astandrik) [inline src/utils/dataFormatters/dataFormatters.ts:79]: please add some condition to prevent infinite loop here

File: src/utils/dataFormatters/dataFormatters.ts [hunk 8]
```diff
@@ -52,7 +52,12 @@ export const formatMsToUptime = (ms?: number) => {
     return ms && formatUptime(ms / 1000);
 };
 
-export const formatStorageValues = (value?: number, total?: number, size?: BytesSizes) => {
+export const formatStorageValues = (
```
Comment (astandrik) [inline src/utils/dataFormatters/dataFormatters.ts:55]: please add jest tests to this function that covers corner cases like infinite loop in while

File: src/utils/numeral.ts [hunk 9]
```diff
@@ -6,7 +6,7 @@ import {Lang, i18n} from './i18n';
 // Set space delimiter for all locales possible in project
 Object.values(Lang).forEach((value) => {
     if (numeral.locales[value]) {
-        numeral.locales[value].delimiters.thousands = ' ';
+        numeral.locales[value].delimiters.thousands = '\xa0';
```
Comment (astandrik) [inline src/utils/numeral.ts:9]: looks like \xa0 should be some global constant 
image

File: src/components/DiagnosticCard/DiagnosticCard.tsx [hunk 10]
```diff
@@ -4,12 +4,20 @@ import './DiagnosticCard.scss';
...
-export function DiagnosticCard({children, className, active}: DiagnosticCardProps) {
-    return <div className={b({active}, className)}>{children}</div>;
+export function DiagnosticCard({
+    children,
+    className,
+    active,
+    size = 'm',
+    interactive = true,
+}: DiagnosticCardProps) {
+    return <div className={b({active, size, interactive}, className)}>{children}</div>;
```
Comment (astandrik) [inline src/components/DiagnosticCard/DiagnosticCard.tsx:22]: Is this card supposed to be used somewhere else in the future
As for now for me it looks like a single-place wrapper for child components without clear self-worthness

Comment (astandrik) [reply src/components/DiagnosticCard/DiagnosticCard.tsx:22]: oh, I see where it is used, sorry

File: src/components/DoughnutMetrics/DoughnutMetrics.tsx (hunk 1 above)
Comment (Raubzeug) [reply src/components/DoughnutMetrics/DoughnutMetrics.tsx:52]: Those styles depends on component's props. It seems using `styles` here is justified...

File: src/containers/Cluster/ClusterDashboard/utils.tsx (hunk 5 above)
Comment (Raubzeug) [reply src/containers/Cluster/ClusterDashboard/utils.tsx:61]: Why? For now there's nothing to localize, and it just adds an abstraction level...

File: src/utils/dataFormatters/dataFormatters.ts (hunk 7 above)
Comment (Raubzeug) [reply src/utils/dataFormatters/dataFormatters.ts:79]: It seems no infinite loops is available here, because before enter the cycle I check if `value` is not nullish.

File: src/utils/utils.ts [hunk 11]
```diff
@@ -104,3 +104,5 @@ export function isNumeric(value?: unknown): value is number | string {
 export function toExponential(value: number, precision?: number) {
     return Number(value).toExponential(precision);
 }
+
+export const unbreakableGap = '\xa0';
```
Comment (astandrik) [inline src/utils/utils.ts:108]: constants are usually named in screaming snake case

**Review by astandrik — APPROVED** [review]

**Review by astandrik — APPROVED** [review]
